ceno_zkvm/src/tables/ram/ram_impl.rs

```diff
@@ -396,46 +396,41 @@ impl<DVRAM: DynVolatileRamTable + Send + Sync + Clone> DynVolatileRamTableConfig
         num_witness: usize,
         final_mem: &[MemFinalRecord],
     ) -> Result<RowMajorMatrix<F>, ZKVMError> {
-        assert!(final_mem.len() <= DVRAM::max_len());
+        let accessed_addrs = final_mem
+            .iter()
+            .cloned()
+            .map(|record| (record.addr, record))
+            .collect::<HashMap<_, _>>();
         assert!(DVRAM::max_len().is_power_of_two());
-        let mut final_table =
-            RowMajorMatrix::<F>::new(final_mem.len().next_power_of_two(), num_witness);
+        let mut final_table = RowMajorMatrix::<F>::new(DVRAM::max_len(), num_witness);
 
         final_table
             .par_iter_mut()
             .with_min_len(MIN_PAR_SIZE)
-            .zip(final_mem.into_par_iter())
-            .for_each(|(row, rec)| {
-                set_val!(row, self.addr, rec.addr as u64);
-                if self.final_v.len() == 1 {
-                    // Assign value directly.
-                    set_val!(row, self.final_v[0], rec.value as u64);
+            .enumerate()
+            .for_each(|(i, row)| {
+                let addr = DVRAM::addr(i);
```

Comment:
I think this should work the other way around -- the tables should take the set of addresses as input rather than impose it and try to reconcile it all that way.

But this PR is progress.